### DISCUSSION TITLE
Cache API Routes

### DISCUSSION THREAD

**Discussion by janhesters** [vercel/next.js#15892] [Help]
I'm looking for a way to cache some data in my [Next.js API route](https://nextjs.org/docs/api-routes/introduction).

I'm creating a route that needs to verify a [Cognito token](https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-using-tokens-verifying-a-jwt.html). As part of this I want to cache two things:

1. The JSON Web keys for a specific user pool and region. I want to cache this for an hour.
1. Whether an access token is valid (so I won't need to rerun all calculations on each request). I want to cache this for a minute.

[Some people](https://github.com/awslabs/aws-support-tools/blob/master/Cognito/decode-verify-jwt/decode-verify-jwt.ts) cache their keys by defining them in an object out of any functions scope.

```ts
let cacheKeys: MapOfKidToPublicKey | undefined;
const getPublicKeys = async (): Promise<MapOfKidToPublicKey> => {
  if (!cacheKeys) {
    const url = `${cognitoIssuer}/.well-known/jwks.json`;
    const publicKeys = await Axios.default.get<PublicKeys>(url);
    cacheKeys = publicKeys.data.keys.reduce((agg, current) => {
      const pem = jwkToPem(current);
      agg[current.kid] = {instance: current, pem};
      return agg;
    }, {} as MapOfKidToPublicKey);
    return cacheKeys;
  } else {
    return cacheKeys;
  }
};
```

But the problem with Next.js functions is that this scope doesn't persist between requests. Furthermore, caching like this doesn't include a timer to invalidate the cache.

I was wondering if there is a way to create a timed cache for Next.js API routes, or any cache at all, so that data is memoized between requests? If yes, how do you do that?

**Comment by stigkj** [comment]
A little late, but you could cache to file.
